session/schema_amender.go

```diff
@@ -0,0 +1,470 @@
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/rowcodec"
+	"go.uber.org/zap"
+)
+
+const amendableType = nonMemAmendType | (memBufAmendType)
+const nonMemAmendType = (1 << model.ActionAddColumn) | (1 << model.ActionDropColumn) | (1 << model.ActionDropIndex)
+const memBufAmendType = uint64(1 << model.ActionAddIndex)
+
+// Amend operation types.
+const (
+	AmendNone int = iota
+
+	// For add index.
+	AmendNeedAddDelete
+	AmendNeedAddDeleteAndInsert
+	AmendNeedAddInsert
+)
+
+// ConstOpAddIndex is the possible ddl state changes, and related amend action types.
+var ConstOpAddIndex = map[model.SchemaState]map[model.SchemaState]int{
+	model.StateNone: {
+		model.StateDeleteOnly:          AmendNeedAddDelete,
+		model.StateWriteOnly:           AmendNeedAddDeleteAndInsert,
+		model.StateWriteReorganization: AmendNeedAddDeleteAndInsert,
+		model.StatePublic:              AmendNeedAddDeleteAndInsert,
+	},
+	model.StateDeleteOnly: {
+		model.StateWriteOnly:           AmendNeedAddInsert,
+		model.StateWriteReorganization: AmendNeedAddInsert,
+		model.StatePublic:              AmendNeedAddInsert,
+	},
+	model.StateWriteOnly: {
+		model.StateWriteReorganization: AmendNone,
+		model.StatePublic:              AmendNone,
+	},
+	model.StateWriteReorganization: {
+		model.StatePublic: AmendNone,
+	},
+}
+
+// amendCollector collects all amend operations, row decoders and memory chunks for each table needs amend.
+type amendCollector struct {
+	tblAmendOpMap map[int64][]amendOp
+	tblDecoder    map[int64]*rowcodec.ChunkDecoder
+	tblChk        map[int64]*chunk.Chunk
+}
+
+func newAmendCollector() *amendCollector {
+	res := &amendCollector{
+		tblAmendOpMap: make(map[int64][]amendOp),
+		tblDecoder:    make(map[int64]*rowcodec.ChunkDecoder),
+		tblChk:        make(map[int64]*chunk.Chunk),
+	}
+	return res
+}
+
+func findIndexByID(tbl table.Table, ID int64) table.Index {
+	for _, indexInfo := range tbl.Indices() {
+		if indexInfo.Meta().ID == ID {
+			return indexInfo
+		}
+	}
+	return nil
+}
+
+func findColByID(tbl table.Table, colID int64) *table.Column {
+	for _, colInfo := range tbl.Cols() {
+		if colInfo.ID == colID {
+			return colInfo
+		}
+	}
+	return nil
+}
+
+func (a *amendCollector) collectIndexAmendOps(ctx context.Context, phyTblID int64,
+	tblAtStart, tblAtCommit table.Table) ([]amendOp, error) {
+	res := make([]amendOp, 0, 4)
+	// Check index having state change, collect index column info.
+	for _, idxInfoAtCommit := range tblAtCommit.Indices() {
+		idxInfoAtStart := findIndexByID(tblAtStart, idxInfoAtCommit.Meta().ID)
+		// Try to find index state change.
+		var amendOpType int
+		if idxInfoAtStart == nil {
+			amendOpType = ConstOpAddIndex[model.StateNone][idxInfoAtCommit.Meta().State]
+		} else if idxInfoAtCommit.Meta().State > idxInfoAtStart.Meta().State {
```

Comment:
I think the State value is not ordered.
Can we just compare with `!=`?

Reply:
New operations will be generated only for `add index`, and states will be [ordered](https://github.com/pingcap/parser/blob/master/model/model.go#L31) for this action type. Other action types like `drop index` will be filtered using greater to compare.